ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,144 @@ PHP_FUNCTION(mb_strtolower)
 	RETURN_STR(mbstring_convert_case(PHP_UNICODE_CASE_LOWER, ZSTR_VAL(str), ZSTR_LEN(str), enc));
 }
 
+typedef enum {
+	MB_LTRIM = 1,
+	MB_RTRIM = 2,
+	MB_BOTH_TRIM = 3
+} MB_TRIM_MODE;
+
+static bool is_trim_wchar(uint32_t w, HashTable *ht)
+{
+	return zend_hash_index_exists(ht, w);
+}
+
+static zend_string* trim_each_wchar(zend_string *str, zend_string *what, MB_TRIM_MODE mode, const mbfl_encoding *enc)
+{
+	unsigned char *in = (unsigned char*)ZSTR_VAL(str);
+	unsigned char *what_in = (unsigned char*)ZSTR_VAL(what);
+	uint32_t wchar_buf[128];
+	uint32_t what_wchar_buf[128];
+	size_t in_len = ZSTR_LEN(str);
+	size_t out_len = 0;
+	size_t what_out_len = 0;
+	size_t what_len = ZSTR_LEN(what);
+	unsigned int state = 0;
+	size_t left = 0;
+	size_t right = 0;
+	size_t total_len = 0;
+	bool first = true;
+
+	if (what_len == 0) {
+		return zend_string_copy(str);
+	}
+
+	HashTable what_ht;
+	zval val;
+	ZVAL_EMPTY_STRING(&val);
+	zend_hash_init(&what_ht, what_len, NULL, NULL, false);
+
+	while (what_len) {
+		what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+		ZEND_ASSERT(what_out_len <= 128);
+		for (size_t i = 0; i < what_out_len; i++) {
+			zend_hash_index_add_new(&what_ht, what_wchar_buf[i], &val);
+		}
+	}
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+		total_len += out_len;
+
+		if (first && (mode & MB_LTRIM)) {
+			for (size_t i = 0; i < out_len; i++) {
+				uint32_t w = wchar_buf[i];
+				if (is_trim_wchar(w, &what_ht)) {
+					left += 1;
+				} else {
+					first = false;
```

Comment:
Trick: you can use `mode ^= MB_LTRIM;` to unset `MB_LTRIM` from `mode`. Then you can get rid of the `first` variable. This saves one comparison.

Reply:
Thank you! I used your XOR code.